spec/formatting.md

```diff
@@ -18,3 +18,236 @@ the local variable takes precedence.
+
+  For example, attempting to format any of the following messages
+  may result in a Formatting error if done within a context that
+
+  1. provides for the variable reference `$user` to resolve to
+     an object `{ name: 'Kat', id: 1234 }`, and
+  2. uses a `:get` formatting function which requires its argument to be an object and
+     an option `field` to be provided with a string value,
+
+  ```
+  {Hello, {(horse) :get field=name}!}
+  ```
+
+  ```
+  {Hello, {$user :get}!}
+  ```
+
+  ```
+  let $id = {$user :get field=id}
+  {Hello, {$id :get field=name}!}
+  ```
+
+Syntax and Data Model errors must be emitted as soon as possible.
+
+During selection, an Expression handler must only emit Resolution and Selection errors.
+During formatting, an Expression handler must only emit Resolution and Formatting errors.
+
+In all cases, when encountering an error,
+a message formatter must provide some representation of the message.
+An informative error or errors must also be separately provided.
+When a message contains more than one error,
+or contains some error which leads to further errors,
+an implementation which does not emit all of the errors
+should prioritise Syntax and Data Model errors over others.
+
+When an error occurs in the resolution of an Expression or Markup Option,
+the Expression or Markup in question is processed as if the option were not defined.
+This may allow for the fallback handling described below to be avoided,
+though an error must still be emitted.
+
+When an error occurs within a Selector,
+the selector must not match any VariantKey other than the catch-all `*`
+and a Resolution or Selector error is emitted.
+
+## Fallback String Representations
+
+The formatted string representation of a message with a Syntax or Data Model error
+is the concatenation of U+007B LEFT CURLY BRACKET `{`,
+a fallback string,
+and U+007D RIGHT CURLY BRACKET `}`.
+If a fallback string is not defined,
+the U+FFFD REPLACEMENT CHARACTER `�` character is used,
+resulting in the string `{�}`.
+
+When an error occurs in a Placeholder that is being formatted,
+the fallback string representation of the Placeholder
+always starts with U+007B LEFT CURLY BRACKET `{`
+and ends with U+007D RIGHT CURLY BRACKET `}`.
+Between the brackets, the following contents are used:
+
+- Expression with Literal Operand: U+0028 LEFT PARENTHESIS `(`
+  followed by the value of the Literal,
+  and then by U+0029 RIGHT PARENTHESIS `)`
+
+  Examples: `{(horse)}`, `{(42)}`
+
+- Expression with Variable Operand: U+0024 DOLLAR SIGN `$`
+  followed by the Variable Name of the Operand
+
+  Example: `{$user}`
+
+- Expression with no Operand: U+003A COLON `:` followed by the Expression Name
+
+  Example: `{:platform}`
+
+- Markup start: U+002B PLUS SIGN `+` followed by the MarkupStart Name
+
+  Example: `{+tag}`
+
+- Markup end: U+002D HYPHEN-MINUS `-` followed by the MarkupEnd Name
+
+  Example: `{-tag}`
+
+- Otherwise: The U+FFFD REPLACEMENT CHARACTER `�` character
```

Comment:
Proposal: change this to "the string between `{` and `}`, as is.
I would assume we get here is none of the cases before was encountered.

So probably a syntax error (bad sigil, for example, or bad format for parameters):
`"Encountered {?count} penguins"` is more useful than `"Encountered {�} penguins"`
Or 
`"Encountered {$count :number invalid options ?? format} penguins"`

Reply:
Yes, this is catching anything not covered by the above. I _think_ that's currently an empty set, as syntax errors use `{�}` or `{fallback string}` for the whole message. So this should never happen, but let's still include it to make sure that we don't end up with undefined behaviour.

I would strongly prefer not including any explicit reference to syntax source here, as that would require keeping a reference to it in all implementations, and might not apply at all if the source is not an MF2 syntax message.